rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -1,94 +1,314 @@
+        size_t size() const {
+            switch (ownership) {
+                case BatchOwnership::Borrowed:
+                    return storage.borrowed.num_instances;
+                case BatchOwnership::VectorOwned:
+                    return storage.vector_owned.size();
+                case BatchOwnership::Moved:
+                    return 0;
+            }
+        }
+
+        /// Serializes the component batch into a rerun datacell that can be sent to a store.
+        Result<SerializedComponentBatch> serialize() const {
+            // TODO(andreas): `to_data_cell` should actually get our storage representation passed
+            // in which we'll allow "type adaptors" in the future (for e.g. setting a stride or
+            // similar).
+            // TODO(andreas): For improved error messages we should add a static_assert that
+            // TComponent implements `to_data_cell`.
+            switch (ownership) {
+                case BatchOwnership::Borrowed: {
+                    auto cell_result = TComponent::to_data_cell(
+                        storage.borrowed.data,
+                        storage.borrowed.num_instances
+                    );
+                    RR_RETURN_NOT_OK(cell_result.error);
+                    return SerializedComponentBatch(
+                        storage.borrowed.num_instances,
+                        std::move(cell_result.value)
+                    );
+                }
+
+                case BatchOwnership::VectorOwned: {
+                    auto cell_result = TComponent::to_data_cell(
+                        storage.vector_owned.data(),
+                        storage.vector_owned.size()
+                    );
+                    RR_RETURN_NOT_OK(cell_result.error);
+                    return SerializedComponentBatch(
+                        storage.vector_owned.size(),
+                        std::move(cell_result.value)
+                    );
+                }
+
+                case BatchOwnership::Moved:
+                    return Error(
+                        ErrorCode::InvalidOperationOnMovedObject,
+                        "ComponentBatch was already moved and is now invalid."
+                    );
+            }
+        }
+
+        /// Returns the ownership of the component batch.
+        ///
+        /// This is usually only needed for debugging and testing.
+        BatchOwnership get_ownership() const {
+            return ownership;
         }
 
       private:
-        Result<rerun::DataCell> (*to_data_cell_func)(const void*, size_t);
+        BatchOwnership ownership;
+
+        template <typename T>
+        union ComponentBatchStorage {
+            struct {
+                const T* data;
+                size_t num_instances;
+            } borrowed;
+
+            std::vector<T> vector_owned;
+
+            ComponentBatchStorage() {}
+
+            ~ComponentBatchStorage() {}
+        };
+
+        ComponentBatchStorage<TComponent> storage;
```

Comment:
Keeping the members this far apart makes it hard to overview.

I think we should also try to use a common style for naming of member variables; either `_storage` or `m_storage`. It really helps reading member functions where it otherwise can be hard to figure out due to the implicit `this->` of C++